tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/RetryExceptionHierarchyTest.java

```diff
@@ -0,0 +1,216 @@
+    }
+
+    @Test
+    public void serviceAthrowsE1() {
+        assertEquals(invokeServiceA(new E1()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsE2() {
+        // I'd rather expect RetryStatus.RETRIED_INVOCATION
+        assertEquals(invokeServiceA(new E2()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsE2S() {
+        // I'd rather expect RetryStatus.RETRIED_INVOCATION
+        assertEquals(invokeServiceA(new E2S()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsE1S() {
+        assertEquals(invokeServiceA(new E1S()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsE0S() {
+        assertEquals(invokeServiceA(new E0S()), RetryStatus.RETRIED_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsRuntimeException() {
+        assertEquals(invokeServiceA(new RuntimeException()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceAthrowsThrowable() {
+        assertEquals(invokeServiceA(new Throwable()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    private RetryStatus invokeServiceA(Throwable exception) {
+        try {
+            service.serviceA(exception);
+        }
+        catch (Throwable expected) {
+        }
+        return service.getStatus();
+    }
+
+    // serviceB
+
+    @Test
+    public void serviceBthrowsException() {
+        assertEquals(invokeserviceB(new Exception()), RetryStatus.RETRIED_INVOCATION);
+    }
+
+    @Test
+    public void serviceBthrowsE0() {
+        assertEquals(invokeserviceB(new E0()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceBthrowsE1() {
+        // I'd rather expect RetryStatus.RETRIED_INVOCATION
+        assertEquals(invokeserviceB(new E1()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceBthrowsE2() {
+        assertEquals(invokeserviceB(new E2()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceBthrowsE2S() {
+        assertEquals(invokeserviceB(new E2S()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceBthrowsE1S() {
+        // I'd rather expect RetryStatus.RETRIED_INVOCATION
+        assertEquals(invokeserviceB(new E1S()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceBthrowsE0S() {
+        assertEquals(invokeserviceB(new E0S()), RetryStatus.FIRST_INVOCATION);
+    }
+
+    @Test
+    public void serviceBthrowsRuntimeException() {
+        assertEquals(invokeserviceB(new RuntimeException()), RetryStatus.RETRIED_INVOCATION);
+    }
+
+    @Test
+    public void serviceBthrowsThrowable() {
+        // I'd rather expect RetryStatus.FIRST_INVOCATION
+        // BUT: this is probably affected by @Retry being specified to ignore custom Throwable-s, which is stupid
+        assertEquals(invokeserviceB(new Throwable()), RetryStatus.RETRIED_INVOCATION);
```

Comment:
I don't think this is right, but I was surprised to find that this passes on openliberty.

In our case, it's because the `Throwable` has been wrapped by a `WeldException` by the time it's caught by the Fault Tolerance interceptor. This is presumably because `InvocationContext.proceed()` is only declared to throw `Exception`.

If I change the test to use `Error` instead of `Throwable` then the test fails as I would expect because no retry occurs.

We should decide whether custom `Throwable` is something that we want to test.

Reply:
Custom `Throwable` is a corner case, but testing corner cases is IMHO what makes a difference :-)

Reply:
True, my only concern is whether it would make it impossible to write a portable implementation. I can't see anything in the Interceptors spec about what happens in an `@AroundInvoke` interceptor if a `Throwable` is thrown from a business method.

Reply:
OK, I had to debug our impl to understand, and same thing happens in our case -- `Throwable` is wrapped into `WeldException` by Weld's `AroundInvokeInvocationContext.proceed`.

I'm looking into chapter 2.5 of the Interceptors 1.2 spec, and it doesn't really answer the question for me.

But we probably don't need to hold this work because of this corner case.

Reply:
So technically, Interceptors spec, version 1.2, chapter 2.5 says:

> Interceptor methods are allowed to throw runtime exceptions or any checked exceptions that the associated target method or constructor allows within its `throws` clause.
>
> ...
>
> The invocation of the `InvocationContext.proceed` method throws the same exception as any thrown by the associated target method ...

And the Java Language Specification, version 11, chapter 11.1.1 says:

> The _unchecked exception_ classes are the run-time exception classes and the error classes.
>
> The _checked exception_ classes are all exception classes other than the unchecked exception classes. That is, the checked exception classes are _Throwable_ and all its subclasses other than _RuntimeException_ and its subclasses and _Error_ and its subclasses.

So `Throwable` is a checked exception, and interceptors should handle it transparently.

At the same time, JLS also says:

> Exception is the superclass of all the exceptions from which ordinary programs may wish to recover.

So I'm not quite sure what to think of it.

Shall we comment the `Throwable` tests out for now? I'm probably making too big of a deal from this :-)

Reply:
Yeah, I'd comment it out for now and we can discuss it on the call.

Reply:
Just out of curiosity, I'm looking at what OpenWebBeans do:

- https://github.com/apache/openwebbeans/blob/master/webbeans-impl/src/main/java/org/apache/webbeans/intercept/InterceptorInvocationContext.java#L51
- https://github.com/apache/openwebbeans/blob/master/webbeans-impl/src/main/java/org/apache/webbeans/intercept/AbstractInvocationContext.java#L118
- https://github.com/apache/openwebbeans/blob/master/webbeans-impl/src/main/java/org/apache/webbeans/util/ExceptionUtil.java#L34

It seems they would rethrow "sneakily" even a `Throwable`, but I didn't try it :-)

In any case, gonna disable these tests for now.